File: 1367.linked-list-in-binary-tree.java
#
# @lc app=leetcode id=1367 lang=java
#
# [1367] Linked List in Binary Tree
#
# @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 */
class Solution {
    public boolean isSubPath(ListNode head, TreeNode root) {
        if (root == null) return false;
        // Start DFS from each node to find a matching path
        return dfs(root, head) || isSubPath(head, root.left) || isSubPath(head, root.right);
    }
    
    private boolean dfs(TreeNode node, ListNode head) {
        if (head == null) return true; // LinkedList fully matched
        if (node == null) return false; // Reached end of path without full match
        if (node.val != head.val) return false; // Mismatch between tree and list nodes
        // Continue DFS on both children with next list node
        return dfs(node.left, head.next) || dfs(node.right, head.next);
    }
}
# @lc code=end

File: 1367.linked-list-in-binary-tree.go
#
# @lc app=leetcode id=1367 lang=golang
#
# [1367] Linked List in Binary Tree
#

# @lc code=start
/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/
/**
* Definition for a binary tree node.
* type TreeNode struct {
*     Val int
*     Left *TreeNode
*     Right *TreeNode
* }
*/
func isSubPath(head *ListNode, root *TreeNode) bool {
    if head == nil { return true } 
    if root == nil { return false } 
    return dfs(head, root) || isSubPath(head, root.Left) || isSubPath(head, root.Right) 
}
func dfs(head *ListNode, node *TreeNode) bool {
    if head == nil { return true } 
    if node == nil { return false } 
    if head.Val != node.Val { return false } 
    return dfs(head.Next, node.Left) || dfs(head.Next, node.Right) 
}
# @lc code=end